parquet-thrift/src/main/java/parquet/thrift/BufferedProtocolReadToWrite.java

```diff
@@ -394,10 +404,42 @@ public String toDebugString() {
           return "f=" + currentField.id + "<t=" + typeName(currentField.type) + ">: ";
         }
       });
-      hasFieldsIgnored |= readOneValue(in, field.type, buffer, expectedField.getType());
+      boolean wasIgnored = readOneValue(in, field.type, buffer, expectedField.getType());
+      if (!wasIgnored) {
+        childFieldsPresent++;
+      }
+      hasFieldsIgnored |= wasIgnored;
       in.readFieldEnd();
       buffer.add(FIELD_END);
     }
+
+    // check that union had exactly 1 (no more no less) child fields.
+    switch (type.getStructOrUnionType()) {
+      case STRUCT:
+        // nothing to do
+        break;
+      case UNION:
+        // childFieldsPresent must == 1
+        if (childFieldsPresent != 1) {
+
+          if (childFieldsPresent == 0) {
+            throw new DecodingSchemaMismatchException("Cannot write a TUnion with no set value in :\n" + type);
+          } else {
+            throw new DecodingSchemaMismatchException("Cannot write a TUnion with more than 1 set value in :\n" + type);
+          }
+
+        }
+        break;
+      case UNKNOWN:
+        // we should never reach here in the write path -- this only happens if the
+        // deprecated constructor of StructType is used, which should only be used in the
+        // read path.
+        throw new ParquetRuntimeException("This should never happen! "
+            + "Don't know if this field is a union, was the deprecated constructor of StructType used?\n" + type){};
+      default:
+        throw new ParquetRuntimeException("Unrecognized StructOrUnionType: " + type.getStructOrUnionType()){};
+    }
```

Comment:
we can do that if we don't worry about being exhaustive and throwing when unrecognized enum members are encountered. I didn't make this class follow the visitor pattern, so we can choose to do this, but if we ever add another type to this enum we may forget to update this line.